File: app/frontend/services/purchaseOrderService.ts
// API configuration
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

// Types
export interface Supplier {
  id: string;
  name: string;
  company_name?: string;
  contact_person?: string;
  email?: string;
  phone?: string;
  address?: string;
  payment_terms?: string;
  lead_time_days: number;
  minimum_order_value: number;
  rating: number;
  total_orders: number;
  on_time_delivery_rate: number;
  is_active: boolean;
  is_preferred: boolean;
  created_at: string;
  updated_at: string;
}

export interface PurchaseOrderItem {
  id?: string;
  product_id: string;
  quantity_ordered: number;
  quantity_received?: number;
  unit_price: number;
  total_price?: number;
  product_name?: string;
  product_sku?: string;
  supplier_sku?: string;
  is_received?: boolean;
  is_quality_checked?: boolean;
  quality_notes?: string;
  received_date?: string;
}

export interface PurchaseOrder {
  id: string;
  order_number: string;
  supplier_id: string;
  status: 'DRAFT' | 'PENDING' | 'APPROVED' | 'ORDERED' | 'SHIPPED' | 'DELIVERED' | 'CANCELLED';
  order_date: string;
  expected_delivery_date?: string;
  actual_delivery_date?: string;
  subtotal: number;
  tax_amount: number;
  shipping_cost: number;
  discount_amount: number;
  total_amount: number;
  delivery_address?: string;
  delivery_instructions?: string;
  tracking_number?: string;
  reference_number?: string;
  notes?: string;
  created_by: string;
  approved_by?: string;
  approved_at?: string;
  created_at: string;
  updated_at: string;
  supplier: Supplier;
  items: PurchaseOrderItem[];
}

export interface CreatePurchaseOrderData {
  supplier_id: string;
  expected_delivery_date?: string;
  delivery_address?: string;
  delivery_instructions?: string;
  notes?: string;
  items: {
    product_id: string;
    quantity_ordered: number;
    unit_price: number;
    supplier_sku?: string;
  }[];
}

export interface DeliveryTracking {
  id: string;
  purchase_order_id: string;
  tracking_number?: string;
  carrier?: string;
  status: 'PENDING' | 'PICKED_UP' | 'IN_TRANSIT' | 'OUT_FOR_DELIVERY' | 'DELIVERED' | 'FAILED';
  shipped_date?: string;
  estimated_delivery_date?: string;
  actual_delivery_date?: string;
  current_location?: string;
  origin_location?: string;
  destination_location?: string;
  delivered_to?: string;
  delivery_signature?: string;
  delivery_photo_url?: string;
  delivery_notes?: string;
  last_status_update: string;
  status_history?: string;
  created_at: string;
  updated_at: string;
}

export interface PurchaseOrderSummary {
  total_orders: number;
  pending_orders: number;
  shipped_orders: number;
  delivered_orders: number;
  total_value: number;
  pending_value: number;
}

export interface DeliveryMetrics {
  in_transit_count: number;
  delivered_today: number;
  delayed_deliveries: number;
  average_delivery_time: number;
}

export interface ReorderSuggestion {
  product_id: string;
  product_name: string;
  product_sku: string;
  current_stock: number;
  min_threshold: number;
  suggested_quantity: number;
  supplier?: {
    id: string;
    name: string;
    price: number;
    minimum_order: number;
    lead_time: number;
  };
}

// Helper function to get auth token
const getAuthToken = (): string | null => {
  if (typeof window !== 'undefined') {
    return localStorage.getItem('access_token');
  }
  return null;
};

// Helper function to get current user ID
export const getCurrentUserId = (): string => {
  if (typeof window !== 'undefined') {
    // Try to get user from localStorage first
    const userStr = localStorage.getItem('user');
    if (userStr) {
      try {
        const user = JSON.parse(userStr);
        if (user.id) return user.id;
        if (user.email) return user.email; // Fallback to email as ID
      } catch {
        console.warn('Failed to parse user from localStorage');
      }
    }
    
    // Try to decode JWT token to get user info
    const token = getAuthToken();
    if (token) {
      try {
        // Decode JWT payload (simple base64 decode, not verification)
        const payload = JSON.parse(atob(token.split('.')[1]));
        if (payload.sub) return payload.sub; // 'sub' is standard JWT claim for user ID
        if (payload.user_id) return payload.user_id;
        if (payload.email) return payload.email;
      } catch {
        console.warn('Failed to decode JWT token');
      }
    }
  }
  
  // Fallback to a default user ID
  return 'anonymous-user';
};

// Import session manager
import { triggerSessionExpired } from '../hooks/useSessionManager';

// Helper function to make API requests
const apiRequest = async <T>(
  endpoint: string, 
  options: RequestInit = {}
): Promise<T> => {
  const url = `${API_BASE_URL}${endpoint}`;
  const token = getAuthToken();

  console.log('🔍 API Request:', { url, method: options.method || 'GET' });

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    ...(options.headers as Record<string, string> || {}),
  };

  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }

  const config: RequestInit = {
    headers,
    ...options,
  };

  try {
    const response = await fetch(url, config);
    
    if (!response.ok) {
      if (response.status === 401) {
        console.warn('Authentication failed - showing session expired dialog');
        // Trigger the session expired dialog instead of throwing error
        triggerSessionExpired();
        // Return a rejected promise to stop execution
        return Promise.reject(new Error('Session expired'));
      }
      
      const errorData = await response.json().catch(() => ({}));
      console.error('API Error Response:', errorData);
      
      let errorMessage = `HTTP error! status: ${response.status}`;
      if (errorData.detail) {
        if (typeof errorData.detail === 'string') {
          errorMessage = errorData.detail;
        } else if (Array.isArray(errorData.detail)) {
          const validationErrors = errorData.detail.map((err: { loc: string[]; msg: string }) => 
            `${err.loc.join('.')}: ${err.msg}`
          ).join(', ');
          errorMessage = `Validation errors: ${validationErrors}`;
        } else {
          errorMessage = JSON.stringify(errorData.detail);
        }
      } else if (errorData.message) {
        errorMessage = errorData.message;
      }
      
      throw new Error(errorMessage);
    }

    return await response.json();
  } catch (error) {
    console.error(`API request failed: ${endpoint}`, error);
    
    // Enhanced error handling for network issues
    if (error instanceof TypeError && error.message.includes('Failed to fetch')) {
      console.error('Network Error Details:', {
        url,
        endpoint,
        API_BASE_URL,
        message: 'Backend server might not be running or CORS issue'
      });
      throw new Error(`Network Error: Cannot connect to backend server at ${API_BASE_URL}. Please ensure the backend server is running and accessible.`);
    }
    
    throw error;
  }
};

// Purchase Order Service
export class PurchaseOrderService {
  // Supplier methods
  static async getSuppliers(activeOnly: boolean = true): Promise<Supplier[]> {
    const params = new URLSearchParams();
    if (activeOnly) params.append('active_only', 'true');
    
    return apiRequest<Supplier[]>(`/api/purchase-orders/suppliers?${params.toString()}`);
  }

  static async createSupplier(supplierData: Omit<Supplier, 'id' | 'rating' | 'total_orders' | 'on_time_delivery_rate' | 'is_active' | 'is_preferred' | 'created_at' | 'updated_at'>): Promise<Supplier> {
    return apiRequest<Supplier>('/api/purchase-orders/suppliers', {
      method: 'POST',
      body: JSON.stringify(supplierData),
    });
  }

  static async getSupplier(supplierId: string): Promise<Supplier> {
    return apiRequest<Supplier>(`/api/purchase-orders/suppliers/${supplierId}`);
  }

  // Purchase Order methods
  static async getPurchaseOrders(filters: {
    status?: string;
    supplier_id?: string;
    page?: number;
    per_page?: number;
  } = {}): Promise<PurchaseOrder[]> {
    const params = new URLSearchParams();
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        params.append(key, value.toString());
      }
    });

    const result = await apiRequest<PurchaseOrder[]>(`/api/purchase-orders?${params.toString()}`);
    console.log('🔍 Purchase Orders Response:', result);
    
    // Debug: Log the first order to see its structure
    if (result && result.length > 0) {
      console.log('🔍 First Order Structure:', {
        id: result[0].id,
        order_number: result[0].order_number,
        has_items: !!result[0].items,
        items_length: result[0].items?.length,
        has_supplier: !!result[0].supplier,
        supplier_name: result[0].supplier?.name
      });
    }
    
    return result;
  }

  static async createPurchaseOrder(orderData: CreatePurchaseOrderData, createdBy: string): Promise<PurchaseOrder> {
    console.log('🔍 Debug - createdBy value:', createdBy);
    console.log('🔍 Debug - createdBy type:', typeof createdBy);
    console.log('🔍 Debug - createdBy length:', createdBy?.length);
    
    if (!createdBy || createdBy.trim() === '') {
      throw new Error('User ID is required but not found. Please log in again.');
    }
    
    const params = new URLSearchParams();
    params.append('created_by', createdBy);
    
    const url = `/api/purchase-orders?${params.toString()}`;
    console.log('🔍 Debug - Final URL:', url);
    
    return apiRequest<PurchaseOrder>(url, {
      method: 'POST',
      body: JSON.stringify(orderData),
    });
  }

  static async getPurchaseOrder(orderId: string): Promise<PurchaseOrder> {
    return apiRequest<PurchaseOrder>(`/api/purchase-orders/${orderId}`);
  }

  static async updatePurchaseOrder(orderId: string, updates: Partial<PurchaseOrder>): Promise<PurchaseOrder> {
    return apiRequest<PurchaseOrder>(`/api/purchase-orders/${orderId}`, {
      method: 'PUT',
      body: JSON.stringify(updates),
    });
  }

  static async approvePurchaseOrder(orderId: string, approvedBy: string): Promise<{ message: string; success: boolean }> {
    const params = new URLSearchParams();
    params.append('approved_by', approvedBy);
    
    return apiRequest<{ message: string; success: boolean }>(`/api/purchase-orders/${orderId}/approve?${params.toString()}`, {
      method: 'POST',
    });
  }

  static async receiveItems(orderId: string, receiveData: {
    items: Array<{
      item_id: string;
      quantity_received: number;
      quality_notes?: string;
    }>;
    received_by: string;
    notes?: string;
  }): Promise<{ message: string; success: boolean; items_processed: number }> {
    return apiRequest<{ message: string; success: boolean; items_processed: number }>(`/api/purchase-orders/${orderId}/receive`, {
      method: 'POST',
      body: JSON.stringify(receiveData),
    });
  }

  // Delivery Tracking methods
  static async getDeliveryTracking(orderId: string): Promise<DeliveryTracking> {
    return apiRequest<DeliveryTracking>(`/api/purchase-orders/${orderId}/tracking`);
  }

  static async updateDeliveryTracking(orderId: string, trackingData: Partial<DeliveryTracking>): Promise<DeliveryTracking> {
    return apiRequest<DeliveryTracking>(`/api/purchase-orders/${orderId}/tracking`, {
      method: 'PUT',
      body: JSON.stringify(trackingData),
    });
  }

  // Dashboard/Summary methods
  static async getPurchaseOrderSummary(): Promise<PurchaseOrderSummary> {
    return apiRequest<PurchaseOrderSummary>('/api/purchase-orders/summary/orders');
  }

  static async getDeliveryMetrics(): Promise<DeliveryMetrics> {
    return apiRequest<DeliveryMetrics>('/api/purchase-orders/summary/deliveries');
  }

  static async getReorderSuggestions(limit: number = 10): Promise<{ suggestions: ReorderSuggestion[]; total_count: number }> {
    return apiRequest<{ suggestions: ReorderSuggestion[]; total_count: number }>(`/api/purchase-orders/products/reorder-suggestions?limit=${limit}`);
  }
}

// Success/Error message helpers
export const showSuccessMessage = (message: string) => {
  console.log('✅ Success:', message);
};

export const showErrorMessage = (error: unknown) => {
  const message = error instanceof Error ? error.message : 'An unexpected error occurred';
  console.error('❌ Error:', message);
}; 

File: app/backend/services/purchase_order_service.py
from sqlalchemy.orm import Session, joinedload
from sqlalchemy import and_, or_, desc, func
from models.purchase_order import (
    PurchaseOrder, PurchaseOrderItem, SupplierProduct, 
    DeliveryTracking, PurchaseOrderStatus, DeliveryStatus
)
from models.product import Product, Supplier
from models.purchase_schemas import (
    PurchaseOrderCreate, PurchaseOrderUpdate, SupplierCreate,
    DeliveryTrackingUpdate, PurchaseOrderSummary, DeliveryMetrics,
    ReceiveItemsRequest
)
from services.product_service import ProductService
from datetime import datetime, timedelta
import uuid
import json

class PurchaseOrderService:
    def __init__(self, db: Session):
        self.db = db

    # Supplier Methods
    def get_suppliers(self, active_only: bool = True):
        """Get all suppliers"""
        query = self.db.query(Supplier)
        if active_only:
            query = query.filter(Supplier.is_active == True)
        return query.order_by(Supplier.name).all()

    def create_supplier(self, supplier_data: SupplierCreate):
        """Create a new supplier"""
        supplier = Supplier(
            id=str(uuid.uuid4()),
            **supplier_data.dict()
        )
        self.db.add(supplier)
        self.db.commit()
        self.db.refresh(supplier)
        return supplier

    def get_supplier_by_id(self, supplier_id: str):
        """Get supplier by ID"""
        return self.db.query(Supplier).filter(Supplier.id == supplier_id).first()

    # Purchase Order Methods
    def get_purchase_orders(self, status=None, supplier_id=None, page=1, per_page=20):
        """Get purchase orders with filtering"""
        query = self.db.query(PurchaseOrder).options(
            joinedload(PurchaseOrder.order_items),
            joinedload(PurchaseOrder.supplier)
        )
        
        if status:
            query = query.filter(PurchaseOrder.status == status)
        if supplier_id:
            query = query.filter(PurchaseOrder.supplier_id == supplier_id)
        
        offset = (page - 1) * per_page
        return query.order_by(desc(PurchaseOrder.created_at)).offset(offset).limit(per_page).all()

    def create_purchase_order(self, order_data: PurchaseOrderCreate, created_by: str):
        """Create a new purchase order"""
        # Generate order number
        order_number = self._generate_order_number()
        
        # Create purchase order
        purchase_order = PurchaseOrder(
            id=str(uuid.uuid4()),
            order_number=order_number,
            supplier_id=order_data.supplier_id,
            expected_delivery_date=order_data.expected_delivery_date,
            delivery_address=order_data.delivery_address,
            delivery_instructions=order_data.delivery_instructions,
            notes=order_data.notes,
            created_by=created_by,
            status=PurchaseOrderStatus.DRAFT
        )
        
        self.db.add(purchase_order)
        self.db.flush()  # Get the ID without committing
        
        # Create order items and calculate totals
        subtotal = 0.0
        for item_data in order_data.items:
            # Get product details
            product = self.db.query(Product).filter(Product.id == item_data.product_id).first()
            if not product:
                raise ValueError(f"Product with ID {item_data.product_id} not found")
            
            total_price = item_data.quantity_ordered * item_data.unit_price
            subtotal += total_price
            
            order_item = PurchaseOrderItem(
                id=str(uuid.uuid4()),
                purchase_order_id=purchase_order.id,
                product_id=item_data.product_id,
                quantity_ordered=item_data.quantity_ordered,
                unit_price=item_data.unit_price,
                total_price=total_price,
                product_name=product.name,
                product_sku=product.sku,
                supplier_sku=item_data.supplier_sku
            )
            self.db.add(order_item)
        
        # Calculate totals
        tax_amount = subtotal * 0.2  # 20% VAT for UK
        purchase_order.subtotal = subtotal
        purchase_order.tax_amount = tax_amount
        purchase_order.total_amount = subtotal + tax_amount + (purchase_order.shipping_cost or 0)
        
        # Create delivery tracking
        delivery_tracking = DeliveryTracking(
            id=str(uuid.uuid4()),
            purchase_order_id=purchase_order.id,
            status=DeliveryStatus.PENDING
        )
        self.db.add(delivery_tracking)
        
        self.db.commit()
        self.db.refresh(purchase_order)
        return purchase_order

    def get_purchase_order_by_id(self, order_id: str):
        """Get purchase order by ID with all related data"""
        return self.db.query(PurchaseOrder).options(
            joinedload(PurchaseOrder.order_items),
            joinedload(PurchaseOrder.supplier)
        ).filter(PurchaseOrder.id == order_id).first()

    def update_purchase_order(self, order_id: str, update_data: PurchaseOrderUpdate):
        """Update purchase order"""
        purchase_order = self.get_purchase_order_by_id(order_id)
        if not purchase_order:
            return None
        
        update_dict = update_data.dict(exclude_unset=True)
        for field, value in update_dict.items():
            setattr(purchase_order, field, value)
        
        purchase_order.updated_at = datetime.utcnow()
        self.db.commit()
        self.db.refresh(purchase_order)
        return purchase_order

    def approve_purchase_order(self, order_id: str, approved_by: str):
        """Approve a purchase order"""
        purchase_order = self.get_purchase_order_by_id(order_id)
        if not purchase_order:
            return None
        
        purchase_order.status = PurchaseOrderStatus.APPROVED
        purchase_order.approved_by = approved_by
        purchase_order.approved_at = datetime.utcnow()
        purchase_order.updated_at = datetime.utcnow()
        
        self.db.commit()
        return True

    def receive_items(self, order_id: str, receive_data: ReceiveItemsRequest):
        """Receive items and automatically update inventory"""
        purchase_order = self.get_purchase_order_by_id(order_id)
        if not purchase_order:
            return None
        
        product_service = ProductService(self.db)
        received_items = 0
        
        for item_data in receive_data.items:
            # Find the order item
            order_item = self.db.query(PurchaseOrderItem).filter(
                and_(
                    PurchaseOrderItem.purchase_order_id == order_id,
                    PurchaseOrderItem.id == item_data["item_id"]
                )
            ).first()
            
            if order_item:
                # Update order item
                quantity_received = item_data.get("quantity_received", 0)
                order_item.quantity_received += quantity_received
                order_item.is_received = order_item.quantity_received >= order_item.quantity_ordered
                order_item.quality_notes = item_data.get("quality_notes")
                order_item.received_date = datetime.utcnow()
                
                # Update product inventory
                product = self.db.query(Product).filter(Product.id == order_item.product_id).first()
                if product:
                    product.quantity_in_stock += quantity_received
                    product.updated_at = datetime.utcnow()
                
                received_items += 1
        
        # Check if all items are received
        all_items_received = all(
            item.is_received for item in purchase_order.order_items
        )
        
        if all_items_received:
            purchase_order.status = PurchaseOrderStatus.DELIVERED
            purchase_order.actual_delivery_date = datetime.utcnow()
            
            # Update delivery tracking
            delivery_tracking = self.db.query(DeliveryTracking).filter(
                DeliveryTracking.purchase_order_id == order_id
            ).first()
            if delivery_tracking:
                delivery_tracking.status = DeliveryStatus.DELIVERED
                delivery_tracking.actual_delivery_date = datetime.utcnow()
                delivery_tracking.delivered_to = receive_data.received_by
        
        self.db.commit()
        return True

    # Delivery Tracking Methods
    def get_delivery_tracking(self, order_id: str):
        """Get delivery tracking for a purchase order"""
        return self.db.query(DeliveryTracking).filter(
            DeliveryTracking.purchase_order_id == order_id
        ).first()

    def update_delivery_tracking(self, order_id: str, tracking_data: DeliveryTrackingUpdate):
        """Update delivery tracking"""
        tracking = self.get_delivery_tracking(order_id)
        if not tracking:
            return None
        
        update_dict = tracking_data.dict(exclude_unset=True)
        for field, value in update_dict.items():
            setattr(tracking, field, value)
        
        tracking.last_status_update = datetime.utcnow()
        tracking.updated_at = datetime.utcnow()
        
        # Update purchase order status based on delivery status
        purchase_order = self.get_purchase_order_by_id(order_id)
        if purchase_order and tracking.status == DeliveryStatus.DELIVERED:
            purchase_order.status = PurchaseOrderStatus.DELIVERED
            purchase_order.actual_delivery_date = tracking.actual_delivery_date
        elif purchase_order and tracking.status == DeliveryStatus.IN_TRANSIT:
            purchase_order.status = PurchaseOrderStatus.SHIPPED
        
        self.db.commit()
        self.db.refresh(tracking)
        return tracking

    # Dashboard/Summary Methods
    def get_purchase_order_summary(self):
        """Get purchase order summary for dashboard"""
        total_orders = self.db.query(PurchaseOrder).count()
        pending_orders = self.db.query(PurchaseOrder).filter(
            PurchaseOrder.status.in_([PurchaseOrderStatus.PENDING, PurchaseOrderStatus.APPROVED])
        ).count()
        shipped_orders = self.db.query(PurchaseOrder).filter(
            PurchaseOrder.status == PurchaseOrderStatus.SHIPPED
        ).count()
        delivered_orders = self.db.query(PurchaseOrder).filter(
            PurchaseOrder.status == PurchaseOrderStatus.DELIVERED
        ).count()
        
        # Calculate total value
        total_value_result = self.db.query(
            func.sum(PurchaseOrder.total_amount)
        ).scalar() or 0.0
        
        pending_value_result = self.db.query(
            func.sum(PurchaseOrder.total_amount)
        ).filter(
            PurchaseOrder.status.in_([PurchaseOrderStatus.PENDING, PurchaseOrderStatus.APPROVED])
        ).scalar() or 0.0
        
        return PurchaseOrderSummary(
            total_orders=total_orders,
            pending_orders=pending_orders,
            shipped_orders=shipped_orders,
            delivered_orders=delivered_orders,
            total_value=float(total_value_result),
            pending_value=float(pending_value_result)
        )

    def get_delivery_metrics(self):
        """Get delivery metrics for dashboard"""
        in_transit_count = self.db.query(DeliveryTracking).filter(
            DeliveryTracking.status.in_([DeliveryStatus.IN_TRANSIT, DeliveryStatus.OUT_FOR_DELIVERY])
        ).count()
        
        # Delivered today
        today = datetime.utcnow().date()
        delivered_today = self.db.query(DeliveryTracking).filter(
            and_(
                DeliveryTracking.status == DeliveryStatus.DELIVERED,
                func.date(DeliveryTracking.actual_delivery_date) == today
            )
        ).count()
        
        # Delayed deliveries (past expected delivery date)
        delayed_deliveries = self.db.query(PurchaseOrder).filter(
            and_(
                PurchaseOrder.expected_delivery_date < datetime.utcnow(),
                PurchaseOrder.status.in_([
                    PurchaseOrderStatus.APPROVED, 
                    PurchaseOrderStatus.ORDERED, 
                    PurchaseOrderStatus.SHIPPED
                ])
            )
        ).count()
        
        # Calculate average delivery time
        delivered_orders = self.db.query(PurchaseOrder).filter(
            and_(
                PurchaseOrder.status == PurchaseOrderStatus.DELIVERED,
                PurchaseOrder.actual_delivery_date.isnot(None)
            )
        ).all()
        
        if delivered_orders:
            total_delivery_days = sum(
                (order.actual_delivery_date - order.order_date).days
                for order in delivered_orders
            )
            average_delivery_time = total_delivery_days / len(delivered_orders)
        else:
            average_delivery_time = 0.0
        
        return DeliveryMetrics(
            in_transit_count=in_transit_count,
            delivered_today=delivered_today,
            delayed_deliveries=delayed_deliveries,
            average_delivery_time=average_delivery_time
        )

    def get_reorder_suggestions(self, limit: int = 10):
        """Get products that need reordering based on stock levels"""
        low_stock_products = self.db.query(Product).filter(
            Product.quantity_in_stock <= Product.min_stock_threshold,
            Product.is_active == True
        ).limit(limit).all()
        
        suggestions = []
        for product in low_stock_products:
            # Find preferred supplier for this product
            supplier_product = self.db.query(SupplierProduct).filter(
                and_(
                    SupplierProduct.product_id == product.id,
                    SupplierProduct.is_available == True
                )
            ).order_by(SupplierProduct.is_preferred.desc()).first()
            
            suggestion = {
                "product_id": product.id,
                "product_name": product.name,
                "product_sku": product.sku,
                "current_stock": product.quantity_in_stock,
                "min_threshold": product.min_stock_threshold,
                "suggested_quantity": max(product.max_stock_threshold or 100, product.min_stock_threshold * 3),
                "supplier": None
            }
            
            if supplier_product:
                supplier = self.get_supplier_by_id(supplier_product.supplier_id)
                suggestion["supplier"] = {
                    "id": supplier.id,
                    "name": supplier.name,
                    "price": supplier_product.supplier_price,
                    "minimum_order": supplier_product.minimum_order_quantity,
                    "lead_time": supplier_product.lead_time_days
                }
            
            suggestions.append(suggestion)
        
        return {"suggestions": suggestions, "total_count": len(suggestions)}

    def _generate_order_number(self):
        """Generate a unique order number"""
        today = datetime.utcnow()
        date_prefix = today.strftime("%Y%m%d")
        
        # Count orders created today
        orders_today = self.db.query(PurchaseOrder).filter(
            func.date(PurchaseOrder.created_at) == today.date()
        ).count()
        
        sequence = str(orders_today + 1).zfill(3)
        return f"PO-{date_prefix}-{sequence}" 